src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs

```diff
@@ -19,45 +20,104 @@ internal class IpcClient
+        public static async Task<IpcMessage> SendMessageAsync(IpcEndpoint endpoint, IpcMessage message, CancellationToken cancellationToken)
         {
-            var stream = endpoint.Connect(ConnectTimeout);
-            Write(stream, message);
-            response = Read(stream);
-            return stream;
+            using IpcResponse response = await SendMessageGetContinuationAsync(endpoint, message, cancellationToken).ConfigureAwait(false);
+            return response.Message;
         }
 
-        private static void Write(Stream stream, byte[] buffer)
+        /// <summary>
+        /// Sends a single DiagnosticsIpc Message to the dotnet process with PID processId.
+        /// </summary>
+        /// <param name="endpoint">An endpoint that provides a diagnostics connection to a runtime instance.</param>
+        /// <param name="message">The DiagnosticsIpc Message to be sent</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>An <see cref="IpcResponse"/> containing the response message and continuation stream.</returns>
+        public static async Task<IpcResponse> SendMessageGetContinuationAsync(IpcEndpoint endpoint, IpcMessage message, CancellationToken cancellationToken)
         {
-            stream.Write(buffer, 0, buffer.Length);
+            Stream stream = null;
+            try
+            {
+                stream = await endpoint.ConnectAsync(cancellationToken).ConfigureAwait(false);
+
+                await WriteAsync(stream, message, cancellationToken).ConfigureAwait(false);
+
+                IpcMessage response = await ReadAsync(stream, cancellationToken).ConfigureAwait(false);
+
+                return new IpcResponse(response, Release(ref stream));
+            }
+            finally
+            {
+                stream?.Dispose();
+            }
         }
 
         private static void Write(Stream stream, IpcMessage message)
         {
-            Write(stream, message.Serialize());
+            byte[] buffer = message.Serialize();
+            stream.Write(buffer, 0, buffer.Length);
+        }
+
+        private static Task WriteAsync(Stream stream, IpcMessage message, CancellationToken cancellationToken)
+        {
+            byte[] buffer = message.Serialize();
+            return stream.WriteAsync(buffer, 0, buffer.Length, cancellationToken);
         }
 
         private static IpcMessage Read(Stream stream)
         {
             return IpcMessage.Parse(stream);
         }
+
+        private static Task<IpcMessage> ReadAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            return IpcMessage.ParseAsync(stream, cancellationToken);
+        }
+
+        private static Stream Release(ref Stream stream1)
+        {
+            Stream intermediate = stream1;
+            stream1 = null;
+            return intermediate;
+        }
```

Comment:
Just to make sure I'm grokking why you're doing this: this is to make it so you can do the `stream?.Dispose()` in the `catch` blocks in case of failure without worrying about accidentally disposing the stream you just handed to the `IpcResponse`?

Reply:
You've understood it completely. In dotnet-monitor tests, we have a [DisposableBox<T>](https://github.com/dotnet/dotnet-monitor/blob/main/src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/DisposableBox.cs) class that does this generically where you can call `T Release()` to remove the value from the box so that it is not disposed.  It's almost like CComPtr but without reference counting. Over time, I see that the "dispose-on-failure-but-not-on-success" pattern is implemented various in various, multiple ways and decided to codify it. Not sure if this is something interesting to add to diagnostics repo or just to leave what I've implemented as-is.